[widclkinfo] Ensure we cancel focus when widget_utils.swipeOn() pushes widget bar off screen

cbkerr opened the pull request:
Fixes bug with widclkinfo interacting with widget_utils.

Expose a ensure_blur in clock_info that widclkinfo can call. Handles animation with events.

The relevant versions are in my app loader: https://cbkerr.github.io/BangleApps/

I've been testing with the watchface Andark.

## Symptoms

Load widget_utils and widclkinfo.

1. The Clock Info of `widclkinfo` stays focused after the widgets autohide, so the selected clock info (`options.menuA`, `options.menuB`) changes when swiping the screen (like when using menus).
2. When the widgets are hidden, tapping in the top left corner focuses the clock info inside widclkinfo.
3. Swiping up to hide the widget bar also changes the selected Clock Info, and if the hidden `widclkinfo` is focused, swiping down to show widgets also changes the selected clock info.

### Minimal reproduction of bug
With useful debug info printed

```js
// development
WIDGETS = [];
widget_utils = require("widget_utils");

//(() => { // don't load if a clock_info was already loaded
// Load the clock infos
let clockInfoItems = require("clock_info").load();
// Add the
let clockInfoMenu = require("clock_info").addInteractive(clockInfoItems, {
  app: "widclkinfo",
  // Add the dimensions we're rendering to here - these are used to detect taps on the clock info area
  x: 0,
  y: 0,
  w: 72,
  h: 24,
  // You can add other information here you want to be passed into 'options' in 'draw'
  // This function draws the info
  draw: (itm, info, options) => {
    // itm: the item containing name/hasRange/etc
    // info: data returned from itm.get() containing text/img/etc
    // options: options passed into addInteractive
    clockInfoInfo = info;
    if (WIDGETS["clkinfo"])
      WIDGETS["clkinfo"].draw(WIDGETS["clkinfo"]);
  }
});
let clockInfoInfo; // when clockInfoMenu.draw is called we set this up

// The actual widget we're displaying
WIDGETS["clkinfo"] = {
  area: "tl",
  width: clockInfoMenu.w,
  draw: function(e) {
    clockInfoMenu.x = e.x;
    clockInfoMenu.y = e.y;
    var o = clockInfoMenu;
    // Clear the background
    g.reset();
    // indicate focus - make background reddish
    //if (clockInfoMenu.focus) g.setBgColor(g.blendColor(g.theme.bg, "#f00", 0.25));
    if (clockInfoMenu.focus) g.setColor("#f00");
    g.clearRect(o.x, o.y, o.x + o.w - 1, o.y + o.h - 1);
    if (clockInfoInfo) {
      var x = o.x;
      if (clockInfoInfo.img) {
        g.drawImage(clockInfoInfo.img, x, o.y); // draw the image
        x += 24;
      }
      var availableWidth = o.x + clockInfoMenu.w - (x + 2);
      g.setFont("6x8:2").setFontAlign(-1, 0);
      if (g.stringWidth(clockInfoInfo.text) > availableWidth)
        g.setFont("6x8:1x2");
      g.drawString(clockInfoInfo.text, x + 2, o.y + 12); // draw the text
    }
  }
};
//})();

Bangle.setUI('clock');
g.clear();
Bangle.loadWidgets();
require("widget_utils").swipeOn();

setInterval(() => {
  g.clear(Bangle.appRect);
  g.setFont("6x8");

  g.drawString("widget_utils.offset: " + widget_utils.offset, 4, 40);
  g.drawString("Bangle.CLKINFO_FOCUS: " + Bangle.CLKINFO_FOCUS, 4, 50);
  g.drawString("clockInfoMenu.y: " + clockInfoMenu.y, 4, 70);
  g.drawString("clockInfoMenu.focus: " + clockInfoMenu.focus, 4, 60);
  g.drawString("menuA: " + clockInfoMenu.menuA, 4, 80);
  g.drawString("menuB: " + clockInfoMenu.menuB, 4, 90);
  //g.drawString(clock_info.clockInfos, 4, 90);
  // g.drawString("cIM.x: " + clockInfoMenu.x, 4, 100);
  //g.drawString("items: " + clockInfoItems, 4, 90)
}, 500);
```



## Expected behavior
When widgets autohide, the Clock Info in `widclkinfo` blurs itself just like tapping outside of its region.
Tapping only selects the Clock Info in `widclkinfo` if it is visible. (when `widget_utils.offset == 0`)


## To test these fixes
1. Upload to storage the slightly tweaked clock_info to `clock_info`
2. Run this from web IDE:
 
<details>
<summary>Code to upload to RAM from Web IDE that writes custom widget_utils to storage</summary>

```js
const clock_info = require("clock_info"); // only slight modification to expose force_blur

require("Storage").write("mywidut", `
exports.offset = 0;
exports.hide = function() {
  exports.cleanup();
  if (!global.WIDGETS) return;
  g.reset(); // reset colors
  for (var w of global.WIDGETS) {
    if (w._draw) return; // already hidden
    w._draw = w.draw;
    w.draw = () => {};
    w._area = w.area;
    w.area = "";
    if (w.x!=undefined) g.clearRect(w.x,w.y,w.x+w.width-1,w.y+23);
  }
};

/// Show any hidden widgets
exports.show = function() {
  exports.cleanup();
  if (!global.WIDGETS) return;
  for (var w of global.WIDGETS) {
    if (!w._draw) return; // not hidden
    w.draw = w._draw;
    w.area = w._area;
    delete w._draw;
    delete w._area;
    w.draw(w);
  }
};

/// Remove anything not needed if the overlay was removed
exports.cleanupOverlay = function() {
  exports.offset = -24;
  Bangle.setLCDOverlay(undefined, {id: "widget_utils"});
  delete exports.autohide;
  delete Bangle.appRect;
  if (exports.animInterval) {
    clearInterval(exports.animInterval);
    delete exports.animInterval;
  }
  if (exports.hideTimeout) {
    clearTimeout(exports.hideTimeout);
    delete exports.hideTimeout;
  }
};

/// Remove any intervals/handlers/etc that we might have added. Does NOT re-show widgets that were hidden
exports.cleanup = function() {
  exports.cleanupOverlay();
  delete exports.offset;
  if (exports.swipeHandler) {
    Bangle.removeListener("swipe", exports.swipeHandler);
    delete exports.swipeHandler;
  }
  if (exports.origDraw) {
    Bangle.drawWidgets = exports.origDraw;
    delete exports.origDraw;
  }
};

/** Put widgets offscreen, and allow them to be swiped
back onscreen with a downwards swipe. Use .show to undo.
First parameter controls automatic hiding time, 0 equals not hiding at all.
Default value is 2000ms until hiding.
Bangle.js 2 only at the moment. On Bangle.js 1 widgets will be hidden permanently.

Note: On Bangle.js 1 is is possible to draw widgets in an offscreen area of the LCD
and use Bangle.setLCDOffset. However we can't detect a downward swipe so how to
actually make this work needs some thought.
*/
exports.swipeOn = function(autohide) {
  if (process.env.HWVERSION!==2) return exports.hide();
  exports.cleanup();
  if (!global.WIDGETS) return;
  exports.autohide=autohide===undefined?2000:autohide;
  /* TODO: maybe when widgets are offscreen we don't even
  store them in an offscreen buffer? */

  // force app rect to be fullscreen
  Bangle.appRect = { x: 0, y: 0, w: g.getWidth(), h: g.getHeight(), x2: g.getWidth()-1, y2: g.getHeight()-1 };
  // setup offscreen graphics for widgets
  let og = Graphics.createArrayBuffer(g.getWidth(),26,16,{msb:true});
  og.theme = g.theme;
  og._reset = og.reset;
  og.reset = function() {
    return this._reset().setColor(g.theme.fg).setBgColor(g.theme.bg);
  };
  og.reset().clearRect(0,0,og.getWidth(),23).fillRect(0,24,og.getWidth(),25);
  let _g = g;
  exports.offset = -24; // where on the screen are we? -24=hidden, 0=full visible

  function queueDraw() {
    const o = exports.offset;
    Bangle.appRect.y = o+24;
    Bangle.appRect.h = 1 + Bangle.appRect.y2 - Bangle.appRect.y;
    if (o>-24) {
      Bangle.setLCDOverlay(og, 0, o, {
        id:"widget_utils",
        remove:()=>{
          require("widget_utils").cleanupOverlay();
        }
      });
    } else {
      Bangle.setLCDOverlay(undefined, {id: "widget_utils"});
    }
  }

  for (var w of global.WIDGETS) if (!w._draw) { // already hidden
    w._draw = w.draw;
    w.draw = function() {
      g=og;
      this._draw(this);
      g=_g;
      if (exports.offset>-24) queueDraw();
    };
    w._area = w.area;
    if (w.area.startsWith("b"))
      w.area = "t"+w.area.substr(1);
  }

  exports.origDraw = Bangle.drawWidgets;
  Bangle.drawWidgets = ()=>{
    g=og;
    exports.origDraw();
    g=_g;
  };

  function anim(dir, callback) {
    if (exports.animInterval) clearInterval(exports.animInterval);
    exports.animInterval = setInterval(function() {
      exports.offset += dir;
      let stop = false;
      if (dir>0 && exports.offset>=0) { // fully down
        stop = true;
        exports.offset = 0;
        exports.emit("shown");
      } else if (dir<0 && exports.offset<-23) { // fully up
        stop = true;
        exports.offset = -24;
        exports.emit("hidden");
      }
      if (stop) {
        clearInterval(exports.animInterval);
        delete exports.animInterval;
        if (callback) callback();
      }
      queueDraw();
    }, 50);
  }
  // On swipe down, animate to show widgets
  exports.swipeHandler = function(lr,ud) {
    if (exports.hideTimeout) {
      clearTimeout(exports.hideTimeout);
      delete exports.hideTimeout;
    }
    let cb;
    if (exports.autohide > 0) cb = function() {
      exports.hideTimeout = setTimeout(function() {
        anim(-4);
      }, exports.autohide);
    };
    if (ud>0 && exports.offset<0) anim(4, cb);
    if (ud<0 && exports.offset>-24) anim(-4);
  };
  Bangle.on("swipe", exports.swipeHandler);
  Bangle.drawWidgets();
};
`);

const widget_utils = require("mywidut"); // only 8 chars allowed in name

// development
WIDGETS = [];


// code from widclkinfo for debugging

//(() => { // don't load if a clock_info was already loaded
  // Load the clock infos
  let clockInfoItems = clock_info.load();

  let wuo = widget_utils.offset;

  let clockInfoMenu = clock_info.addInteractive(clockInfoItems, {
    app: "widclkinfo",
    // Add the dimensions we're rendering to here - these are used to detect taps on the clock info area
    x: 0,
    y: 0, // set offset to initial offset
    w: 72,
    h: 24,
    // You can add other information here you want to be passed into 'options' in 'draw'
    // This function draws the info
    draw: (itm, info, options) => {
      // itm: the item containing name/hasRange/etc
      // info: data returned from itm.get() containing text/img/etc
      // options: options passed into addInteractive
      clockInfoInfo = info;
      //wuo = 0 | widget_utils.offset;
      clockInfoMenu.y = options.y + wuo;
      if (WIDGETS["clkinfo"]) {
        WIDGETS["clkinfo"].draw(WIDGETS["clkinfo"]);
        console.log("Clock Info was updated, thus drawing widget.");
      }
    }
  });

  let clockInfoInfo; // when clockInfoMenu.draw is called we set this up

  // The actual widget we're displaying
  WIDGETS["clkinfo"] = {
    area: "tl",
    width: clockInfoMenu.w,
    draw: function(e) {
      clockInfoMenu.x = e.x;
      wuo = 0 | widget_utils.offset;
      clockInfoMenu.y = e.y + wuo;
      var o = clockInfoMenu;
      // Clear the background
      g.reset();
      // indicate focus
      if (clockInfoMenu.focus) g.setColor("#f00");
      g.clearRect(o.x, o.y, o.x + o.w - 1, o.y + o.h - 1);

      if (clockInfoInfo) {
        var x = o.x;
        if (clockInfoInfo.img) {
          g.drawImage(clockInfoInfo.img, x, o.y); // draw the image
          x += 24;
        }
        var availableWidth = o.x + clockInfoMenu.w - (x + 2);
        g.setFont("6x8:2").setFontAlign(-1, 0);
        if (g.stringWidth(clockInfoInfo.text) > availableWidth)
          g.setFont("6x8:1x2");
        g.drawString(clockInfoInfo.text, x + 2, o.y + 12); // draw the text
      }
    }
  };

  widget_utils.on("hidden", () => {
    console.log("hidden");
    clockInfoMenu.y = -24;
    clockInfoMenu.force_blur(); // needs to be here so it doesn't stay the focused color
    if (clockInfoMenu.focus) {
      //clockInfoMenu.force_blur();
      console.log("Forced blur bc hidden");
    }
  });

  widget_utils.on("shown", () => {
    clockInfoMenu.y = 0;
    console.log("shown");
    if (WIDGETS["clkinfo"]) {
      WIDGETS["clkinfo"].draw(WIDGETS["clkinfo"]);
    }
  });

  // for debug:
//})();



Bangle.setUI('clock');
g.clear();
Bangle.loadWidgets();
widget_utils.swipeOn();


// debug info:
setInterval(() => {
  g.clear(Bangle.appRect);
  g.setFont("6x8");

  g.drawString("widget_utils.offset: " + widget_utils.offset, 4, 40);
  g.drawString("Bangle.CLKINFO_FOCUS: " + Bangle.CLKINFO_FOCUS, 4, 50);
  g.drawString("clockInfoMenu.y: " + clockInfoMenu.y, 4, 70);
  g.drawString("clockInfoMenu.focus: " + clockInfoMenu.focus, 4, 60);
  g.drawString("menuA: " + clockInfoMenu.menuA, 4, 80);
  g.drawString("menuB: " + clockInfoMenu.menuB, 4, 90);
  //g.drawString(clock_info.clockInfos, 4, 90);
  // g.drawString("cIM.x: " + clockInfoMenu.x, 4, 100);
  //g.drawString("items: " + clockInfoItems, 4, 90)
}, 500);

// these work:
//global.clockInfoMenu
//clockInfoMenu.focus

// get list of all added infos:
//clock_info.clockInfos

// Added a function clockInfoMenu.force_blur()
```

</details>

Commit:
clock_info: Expose internal blur function as force_blur

...so an external action like hiding widgets with widclkinfo can
unfocus the clock info.
This always decrements Bangle.CLKINFO_FOCUS,
so in the future we should only call force_blur if we know the Clock
Info is focused. We could provide a different function ensure_blur.

Commit:
widget_utils: Add shown and hidden events

This can help widgets, such as widclkinfo, know when they are on
screen.

Commit:
widclkinfo: work in progress: Be aware of widget_utils.swipeOn

TODO: disable these changes when widget_utils are not being used.

Use new (prototype) widget_utils events to blur and set offscreen the
clock info when the widgets are hidden. This prevents activating or
further interacting with the widclkinfo.

bobrippling commented:
Nice find, and I like your fix too.

> How can we make these components not interfere while keeping them fairly modular?

I like your event approach - one way to avoid having a dependency between the modules is to copy what's done elsewhere and emit (and listen to) the events on `Bangle`, so:
```diff
-exports.emit("shown");
+Bangle.emit("widgets-shown");
```

For the issue with "refcounting" blur, were there some problems with this initial approach you had?

```diff
-clockInfoMenu.force_blur(); // needs to be here so it doesn't stay the focused color
 if (clockInfoMenu.focus) {
-  //clockInfoMenu.force_blur();
+  clockInfoMenu.blur();
   console.log("Forced blur bc hidden");
 }
```

cbkerr commented:
> For the issue with "refcounting" blur, were there some problems with this initial approach you had?

I was having a problem with the widclkinfo sliding back in still in the redish "focused" color. Actually, in the current state with always calling `clockInfoMenu.force_blur()`, it still slides back in redish. Even when I add extra draws on widgets-hidden, it still comes down red. Even with a frequently updating clock_info like HRM.

Not sure if we should make events for the animation steps.

Commit:
Suggestions from review: emit events from Bangle and moving force_blur calls

Commit:
Also emit event when autohiding

cbkerr commented on apps/widclkinfo/widget.js:
```diff
@@ -47,4 +57,20 @@ if (!require("clock_info").loadCount) { // don't load if a clock_info was alread
-}
\ No newline at end of file
+
+  widget_utils.on("hidden", () => {
+    console.log("hidden");
```
I couldn't figure out how to get the offscreen buffer to re-draw in the normal color before showing, but blurring the clock info before the widgets leave the screen works and also makes sense for the UI.


```suggestion
  Bangle.on("widgets-start-hide", () => {
    console.log("starting to hide");
    if (WIDGETS["clkinfo"]) {
      WIDGETS["clkinfo"].draw(WIDGETS["clkinfo"]);
    }
```

Commit:
Listen for widget events from Bangle

bobrippling commented on modules/widget_utils.js:
```diff
@@ -153,12 +156,19 @@ exports.swipeOn = function(autohide) {
     let cb;
     if (exports.autohide > 0) cb = function() {
       exports.hideTimeout = setTimeout(function() {
+        Bangle.emit("widgets-start-hide");
```
Would it work if we moved the `widgets-start-show/hide` into the start of the `anim()` function?

bobrippling commented on modules/widget_utils.js:
```diff
@@ -11,6 +11,7 @@ exports.hide = function() {
     w.area = "";
     if (w.x!=undefined) g.clearRect(w.x,w.y,w.x+w.width-1,w.y+23);
   }
+  // TODO: do we need to emit event here too?
```
Yes I think so, no `start-...`, just `widgets-hidden` right?

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -47,4 +57,20 @@ if (!require("clock_info").loadCount) { // don't load if a clock_info was alread
-}
\ No newline at end of file
+
+  Bangle.on("hidden", () => {
```
Do we want:

```suggestion
  Bangle.on("widgets-hidden", () => {
```

(and similarly below)

cbkerr replied:
Oops yes, I didn't copy over correctly

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -47,4 +57,20 @@ if (!require("clock_info").loadCount) { // don't load if a clock_info was alread
+  Bangle.on("hidden", () => {
+    console.log("hidden");
+    clockInfoMenu.y = -24;
+    if (clockInfoMenu.focus) {
+      clockInfoMenu.force_blur();
```
Because we've checked `focus` above, I think it's safe to just have the `blur` method and avoid the cost of an extra jsvar, wdyt?

cbkerr replied:
I don't think `clockInfoMenu.blur()` is exposed. I chose the name `force_blur` to avoid internal name confusion inside clock_info. Also, clock_info exposes focus as a boolean but internally uses focus as a function, so I wasn't sure how to keep names consistent

bobrippling replied:
Ah I see your problem - I think that's ok to have the internal focus and the exposed one, we could rename the internal one so we present an API that doesn't give away internal details if you like?

bobrippling reviewed (commented):
Cool, yes this looks good!

Commit:
widget_utils: Emit shown and hidden events on show and hide functions

Commit:
widclkinfo: Remove references to widget_utils

Commit:
widclkinfo: cleanup event listeners

Commit:
widclkinfo: move draw function definition.

Key change to function: not setting `clockInfoMenu.y = e.y`

Commit:
widclkinfo: Don't redraw widget if its clock_info is off screen

I think it's okay to not update the clock_info in this case because
they are designed to be on screen.

Commit:
widclkinfo: balance brackets

Commit:
widclkinfo: Make clockInfoMenu part of the widget

Commit:
widget_utils: Move events to anim function

Commit:
widget_utils: format indentation

Commit:
clock_info: Use ensure_blur as public function in options

Commit:
widclkinfo: update version and changelog

Commit:
clock_info: update changelog

Commit:
clock_info: bump version

Commit:
Remove test code comment

Commit:
Note possible off by one error in detecting taps on clock_info

bobrippling commented:
Let me know when you're happy for a re-review :)

Commit:
widget_utils: Work to fix logic error in animation state handling

1) Run exports.offset+=dir only in animation steps using the existing
   stop variable. We check if near animation boundary instead of doing
   the addition and seeing if we overstepped the animation boundary.

   The logic table now has absolute values that mean "nearly done with
   animation". This should mean that, eg, if the user swipes up in the last
   frame of the show animation, the widgets-start-hide event will
   still be triggered.

2) Also, standardize event name to widgets-*

cbkerr commented on modules/widget_utils.js:
```diff
@@ -129,20 +131,40 @@ exports.swipeOn = function(autohide) {
     if (exports.animInterval) clearInterval(exports.animInterval);
     exports.animInterval = setInterval(function() {
-      exports.offset += dir;
+// exports.offset < -23 + |dir|     > -|dir|         otherwise
+//     dir >0     start showing     shown this step  showing
+//         <0     hidden this step  start hiding     hiding
```
There is probably a more efficient way to do these nested conditions

Commit:
widclkinfo: Remove debug line

Commit:
widget_utils: fix missing bracket

Commit:
widget_utils: Fix ambiguous if else

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -1,50 +1,93 @@
+        // info: data returned from itm.get() containing text/img/etc
+        // options: options passed into addInteractive
+        var wi = WIDGETS["clkinfo"];
+        wi.clockInfoInfo = info;
```
Where does this variable come from? (`info`)

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -1,50 +1,93 @@
+      delete this.init;
+      return this;
+    },
+    clockInfoInfo: undefined, // defined during clockInfoMenu.draw()
```
Perhaps we give this a less repeating name

```suggestion
    info: undefined, // defined during clockInfoMenu.draw()
```

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -1,50 +1,93 @@
+    }
+  });
+
+  Bangle.on("widgets-shown", () => {
+    var wi = WIDGETS["clkinfo"];
+    if (wi) {
+      wi.clockInfoMenu.y = 0;
+      wi.draw(wi);
+    }
+  });
```
Is this the same as the callback for `widgets-start-show`?

bobrippling commented on apps/widclkinfo/widget.js:
```diff
@@ -1,50 +1,93 @@
-      // Clear the background
+    area: "tl",
+    width: 0, //this.clockInfoMenu.w,
+    init: function() {
+      this.width = this.clockInfoMenu.w;
+      delete this.init;
+      return this;
+    },
```
Do we need this function? Can we set `width` at the start?

bobrippling commented on apps/clock_info/lib.js:
```diff
@@ -332,6 +332,10 @@ exports.addInteractive = function(menu, options) {
     if (redraw) options.redraw();
   };
+  // better to only call blur when we know it's focused. Could reuse this logic in this file
+  options.ensure_blur = () => {
```
We can have this have the same name to avoid users of the API knowing about internals:

```suggestion
  options.blur = () => {
```

bobrippling commented on modules/widget_utils.js:
```diff
@@ -129,19 +131,41 @@ exports.swipeOn = function(autohide) {
-        stop = true;
-        exports.offset = -24;
+      if (dir > 0) {
+        if (exports.offset >= -dir) {
+          // nearly shown
+          stop = true;
+          exports.offset = 0; // clamp
+          Bangle.emit("widgets-shown");
+        } else if (exports.offset < -23 + dir) {
+          Bangle.emit("widgets-start-show");
+        } else {
+          Bangle.emit("widgets-anim-step");
+        }
```
Do we need these changes? clkinfo is loaded by a lot of apps so it has a high barrier to additional complexity/code. For instance, `widgets-anim-step` isn't used

bobrippling commented on modules/widget_utils.js:
```diff
@@ -129,19 +131,41 @@ exports.swipeOn = function(autohide) {
         clearInterval(exports.animInterval);
         delete exports.animInterval;
-        if (callback) callback();
+        if (callback) {
+          callback();
+        }
```
Let's avoid style and extra changes and keep the PR focused on one thing :)

```suggestion
        if (callback) callback();
```

bobrippling reviewed (commented):
Looks like there's todos / investigation required going off the comments, happy to re-review when resolved :)

gfwilliams commented:
Closing this as no activity in almost a year - and I haven't seen anyone complaining about it as an issue recently
